locals.tf

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  keyring_id   = var.existing_keyring ? data.google_kms_key_ring.key_ring[0].id : google_kms_key_ring.key_ring[0].id
+  key_opts_map = { for opts in var.key_opts : opts.key => opts }
+
+  key_opts_default = {
+    rotation_period  = "86400s"
+    algorithm        = "GOOGLE_SYMMETRIC_ENCRYPTION"
+    protection_level = "SOFTWARE"
+    labels           = {}
+    prevent_destroy  = true
```

Comment:
We should add purpose field

Reply:
@bharathkkb Hello, actually regarding upgrade guide I'm afraid I have never used the original version, so it would complicated.

Reply:
It should just be adding a few sample state mv commands like we did for [our folders module](https://github.com/terraform-google-modules/terraform-google-folders/blob/master/docs/upgrading_to_folders_v3.0.md#:~:text=To%20prevent%20folder%20recreation%2C%20it%20is%20recommended%20to%20migrate%20the%20statefile%20as%20shown%20below) with the for_each upgrade but we can test this out for an upgrade path if you don't have bandwidth.

Also I don't see the purpose field which the current module added support for https://github.com/terraform-google-modules/terraform-google-kms/blob/c28453ccb6ed33e2c9c7a4e610b2775424536d50/main.tf#L32